FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+ *  houskeeping
+ *  thermal control
+ *
+ * Additional files needed for compilation:
+ *  none
+ */
+
+/* - - - - - - Includes - - - - - - */
+// C++ libraries
+
+// Other libraries
+
+// NS2 headers
+#include "../headers/hkUtil.hpp"
+
+
+/* Module Variable Definitions */
+static HousekeepingData hkData[HK_SAMPLES_TO_KEEP] = {}; // array to store housekeeping data
+static int hkIndex = 0;
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - handleHousekeeping - - - - - - *
+ * Usage:
+ *  Handles all housekeeping functions.
+ *  Samples temperature and power data, handles thermal control
+ *  
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void handleHousekeeping()
+{
+    sampleHousekeepingData();
+    setHeater();
+    hkIndex++;
+    if (hkIndex >= HK_SAMPLES_TO_KEEP) { hkIndex = 0; }
+}
+
+/* - - - - - - Module Helper Functions - - - - - - */
+
+/* - - - - - - setHeater - - - - - - *
+ * Usage:
+ *  Toggles heater based on current optics bench temperature
+ *  Turns heater on when optics temp crosses below HEATER_TEMP_LOW
+ *  Turns heater off when optics temp crosses above HEATER_TEMP_HIGH
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void setHeater()
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
+
+/* - - - - - - sampleHousekeepingData - - - - - - *
+ * Usage:
+ *  samples temperature and power data and stores it in hkData,
+ *  checks for out-of-bounds values and logs faults if found
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void sampleHousekeepingData()
+{
+    // read thermistor voltage values
+    float opticsTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_OPTICS_THERM);
+    float analogTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_ANALOG_THERM);
+    float digitalTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_DIGITAL_THERM);
+    
+    // update housekeeping data
+    latestHkSample.opticsTemp = voltageToTemp(opticsTempVoltage); 
+    latestHkSample.analogTemp = voltageToTemp(analogTempVoltage);
+    latestHkSample.digitalTemp = voltageToTemp(digitalTempVoltage);
+    latestHkSample.analogCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_AREG_CURR);
+    latestHkSample.digitalCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_CURR);   
+    latestHkSample.digitalRegPG = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_PG);
+    hkData[hkIndex] = latestHkSample; // update data array
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
+    else if (latestHkSample.opticsTemp < OPTICS_TEMP_MIN_SAFE) { } // optics temp too low
+
+    if (latestHkSample.analogTemp > BOARD_TEMP_MAX_SAFE) { } // analog board temp too high
```

Comment:
If any of these are outside of the condition, a transition to safe mode may be validated.

Reply:
If any of these trigger, a fault will be logged, and the fault will generate a command to enter safe mode.